### PR TITLE
feat(core): improved some hooks

### PR DESCRIPTION
**Pull request by Yehonal**
## Changes Proposed:

WIP

## Issues Addressed:

- Closes 

## Tests Performed:

- 
- 

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Loot/LootMgr.cpp [hunk 1]
```diff
@@ -291,6 +291,9 @@ bool LootStoreItem::Roll(bool rate, Player const* player, Loot& loot, LootStore
 
     sScriptMgr->OnItemRoll(player, this, _chance, loot, store);
 
+    if (chance == 0)
+        return false;
```
Comment (FrancescoBorzi) [inline src/server/game/Loot/LootMgr.cpp:295]:
```suggestion
    if (chance == 0)
    {
        return false;
    }
```

File: src/server/game/Loot/LootMgr.cpp [hunk 2]
```diff
@@ -1207,6 +1210,9 @@ LootStoreItem const* LootTemplate::LootGroup::Roll(Loot& loot, Player const* pla
 
             sScriptMgr->OnItemRoll(player, item, chance, loot, store);
 
+            if (chance == 0)
+                continue;
```
Comment (FrancescoBorzi) [inline src/server/game/Loot/LootMgr.cpp:1214]:
```suggestion
            if (chance == 0)
            {
                continue;
            }
```

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
we've been using this on prod and no related issues have been reported so far, LGTM

File: src/server/game/Scripting/ScriptMgr.cpp [hunk 3]
```diff
@@ -1650,11 +1650,33 @@ void ScriptMgr::OnPlayerRemoveFromBattleground(Player* player, Battleground* bg)
     FOREACH_SCRIPT(PlayerScript)->OnRemoveFromBattleground(player, bg);
 }
 
+bool ScriptMgr::OnBeforeAchievementComplete(Player* player, AchievementEntry const* achievement)
```
Comment (Kitzunu) [inline src/server/game/Scripting/ScriptMgr.cpp:1653]: wouldnt this be OnAchievementInProgress?

Comment (Yehonal) [reply src/server/game/Scripting/ScriptMgr.cpp:1653]: https://www.azerothcore.org/wiki/hooks-script#naming-conventions

The naming convention we should follow from now on implies this structure: On<When><Action>

normally the Action refers to the parent function (in this case AchievementComplete), while the "When" refers in which part of the function the hooks is called. If we follow the ["Single responsibility principle"](https://en.wikipedia.org/wiki/Single-responsibility_principle), it's very unlikely to have more than 2 hooks per each function (OnBefore<action> and OnAfter<action>). However, edge cases always exist and it's fine to break the convention sometimes if we document our hooks properly. However, in this case, it's not needed.

**Comment by Yehonal** [comment]
@FrancescoBorzi this PR was in WIP and I still had to document and provide the changelog for it...I'll open another one.

**Pull request merged**